### ISSUE TITLE
Reflecting on un-marked type 'List<GeoRange>'

### ISSUE THREAD

**Issue by gi097** [dart-lang/reflectable#147]
How can one make reflecting work on a `List`? Of course the reflection annotation needs to be added on every class which needs to support reflecting. However, can this be done for a list? Or how can one get the type `T` of a `List<T>`?

Reflecting using:
```dart
Type listType = reflectable.reflect(k).type.typeArguments[0].runtimeType;
```
gives `Reflecting on un-marked type 'List<GeoRange>'`.

I also tried the following:
```dart
@reflectable
class Bird {
  @reflectable
  List<GeoRange> GeoRanges = [];
}
```

**Comment by eernstg** [comment]
Reflection on lists is a case that creates a couple of difficulties. First, you should note that the type of a list isn't `List<OfAnything>` (`List` is an abstract class):

```dart
Type typeOf<X>() => X;

main() {
  print(<int>[].runtimeType); // Prints `List<int>`.
  print(<int>[].runtimeType == typeOf<List<int>>()); // Prints `false`.
}
```

So even though the `Type` of a list will print as `List` or `List<...>`, it is actually an instance of some implementation specific subclass of `List` (typically a platform specific class, or once in a blue moon it could be a user-written class that implements List). This means that the actual run-time class of the list is likely to be unknown to you, and it may even be private.

If it is private then reflectable (like _any_ other approach using code generation) cannot denote that class outside the declaring library, and we cannot edit 'dart:core'. This means that reflectable cannot find the relevant mirror for such an object, and that's the reason why you get the 'un-marked type' error. Even though we could generate code for a suitable mirror, there is no expression which will be true iff that mirror is the right one to use.

(You might say "but I just want the `List` mirror for any subtype of `List`", but in that case I'll recommend reading [this](https://github.com/dart-lang/reflectable/blob/master/doc/TheDesignOfReflectableCapabilities.md#considerations-around-admitting-subtypes) to see why we don't support that.)

If you _are_ able to denote the actual class then you can associate a given reflector with that class using a `GlobalQuantifyCapability`, as shown [here](https://github.com/dart-lang/test_reflectable/blob/master/test/new_instance_native_test.dart) for `List`, where it's no problem that `List` is abstract because we are using that class to obtain a class mirror and then using that to invoke a constructor.

So in the special case of lists I usually recommend dealing with them directly: If you actually know that it's a list then test the type and use statically typed code to work with it: `if (x is List) print(x[0]);`. This is easy, safe, performant&mdash;and of course irritating because it is a special case, but maybe you can live with it. ;-)

At the end it looks like your main purpose here is to get the actual type argument, and that's another tricky issue. In general, there is no support for extracting the value of a type argument from a given `Type`, nor from a given instance (we can get its `runtimeType`, but then we just have a `Type` again, for which we can't get those type arguments) short of using built-in reflection ('dart:mirrors'). In other words, we cannot generate code that will give you those type arguments, because no such code exists.

Again, as a special casing workaround, you can easily check for having a subtype of a specific type argument when you know that it is a list:

```dart
bool listTypeArgumentIs<X>(List l) => l is List<X>;
```

but that will of course only allow you to detect that any given type (available as a type that you can express, though not necessarily a compile-time constant value because it can be an expression containing type variables from the context), and you can't obtain the precise value of that type argument (not to mention expressing it such that it can be used as a type in Dart).

So, in summary, you're trying to do exactly a couple of those things where generated code (including anything generated by reflectable) hits a wall because there are no primitives of the relevant kind (and that's because it would probably have serious performance implications to support those primitives in general), and my best advice is to drop back to static code whenever possible, and otherwise to do something which _is_ possible. ;-)

**Comment by gi097** [comment]
Thanks a lot 👍

**Comment by gi097** [comment]
I managed to get something working, which is kinda ugly: https://github.com/gi097/json_parser/blob/ba84c4c895b789c593af101a9e7250111a5043f3/lib/json_parser.dart#L69

But it allows me to determine the type of the list and match it with one of the reflectable class' properties.

**Comment by eernstg** [comment]
Don't tell anybody! ;-)
